typings/styled-components.d.ts

```diff
@@ -65,6 +65,15 @@ type ThemedStyledComponentFactories<T> = ThemedStyledComponentFactoriesHTML<T> &
 export interface ThemedBaseStyledInterface<T> extends ThemedStyledComponentFactories<T> {
   <P extends { theme?: T; }>(component: Component<P>): ThemedStyledFunction<P, T, WithOptionalTheme<P, T>>;
   <P>(component: Component<P>): ThemedStyledFunction<P, T>;
+
+  // Html styled component factories
+  <Props extends { theme?: T; }>(
```

Comment:
to support proper html tags here, we need to take the tag as a unit type and use information from `ThemedStyledComponentFactories<T>` to retrieve proper factory type. Something like this (simplified):
```ts
<K extends keyof HTMLTags>(htmlTag: K): ThemedStyledComponentFactoriesHTML<T>[K];
<K extends keyof SVGTags>(svgTag: K): ThemedStyledComponentFactoriesSVG<T>[K];
```

Reply:
just playing about with this and `K` is still a union of all keys of HTMLTags with this style.

I'm not actually sure if this can be achieved without using a giant overload along the following lines.

```ts
interface DomTagStyledFunction {
  (tag: 'div'):  ThemedStyledComponentFactoriesHTML<T>['div']
```

I'm going to open up a pr with something in glamorous (which has this same issue) and will can redo a pr with whatever style ends up landing there.